pkg/kubelet/volumemanager/reconciler/reconciler.go

```diff
@@ -251,77 +250,302 @@ func (rc *reconciler) reconciliationLoopFunc() func() {
-							glog.Errorf(
-								"operationExecutor.DetachVolume failed for volume %q (spec.Name: %q) controllerAttachDetachEnabled: %v with err: %v",
-								attachedVolume.VolumeName,
-								attachedVolume.VolumeSpec.Name(),
-								rc.controllerAttachDetachEnabled,
-								err)
-						}
-						if err == nil {
-							glog.Infof("DetachVolume operation started for volume %q (spec.Name: %q)",
-								attachedVolume.VolumeName,
-								attachedVolume.VolumeSpec.Name())
-						}
-					}
 				}
 			}
 		}
 	}
 }
+
+// reconstruc process tries to observes the real world by scaning all pods' volumes directories from the disk.
+// If the actual and desired state of worlds are not consistent with the observed world, that means tsome
+// mounted volumes are left out probably during kubelet restart. This process will reconstruct
+// the volumes and udpate the actual and desired states. In the following reconciler loop, those volumes will
+// be cleaned up.
+func (rc *reconciler) reconstruct() {
+	glog.V(1).Infof("Start the routine to reconstruct states by reading volumes dirs from disk directories %q", rc.kubeletPodsDir)
+	defer rc.updateReconstructTime()
+	rc.reconstructStates(rc.kubeletPodsDir)
+}
+
+func (rc *reconciler) updateReconstructTime() {
+	rc.timeOfLastReconstruct = time.Now()
+}
+
+type podVolume struct {
+	podName        volumetypes.UniquePodName
+	volumeSpecName string
+	mountPath      string
+	pluginName     string
+}
+
+// reconstructFromDisk scans the volume directories under the given pod directory. If the volume is not
+// in either actual or desired states of world, or pending operation, this function will reconstuct
+// the volume spec and put it in both the actual and desired state of worlds. If no running
+// container is mounting the volumes, they will be removed by desired state of world's populator and
+// got cleaned up by the reconciler.
+func (rc *reconciler) reconstructStates(podsDir string) {
+	// Get volumes information by reading the pod's directory
+	podVolumes, err := getVolumesFromPodDir(podsDir)
+	if err != nil {
+		glog.Errorf("Cannot get volumes from disk %v", err)
+		return
+	}
+	for _, volume := range podVolumes {
+		volumeToMount, err := rc.reconstructVolume(volume)
+		if err != nil {
+			glog.Errorf("Could not construct volume information: %v", err)
+			continue
+		}
+
+		// Check if there is an pending operation for the given pod and volume.
+		// Need to check pending operation before checking the actual and desired
+		// states to avoid race condition during checking. For exmaple, the following
+		// might happen if pending operation is checked after two states checking.
+		// 1. checking pod does not exist in both actual and desired states.
+		// 2. An operation for the given pod just finishes and the actual state is updated.
+		// 3. checking no pending operation for the given pod.
+		if rc.operationExecutor.IsOperationPending(volumeToMount.VolumeName, volumeToMount.PodName) {
```

Comment:
@saad-ali @jingxu97 Is this worth logging for later? Would we want to know this incase we are debugging?